src/elastio-snap.c

```diff
@@ -1118,6 +1119,17 @@ static inline void tracer_set_fail_state(struct snap_device *dev, int error){
 	smp_mb();
 }
 
+static inline int tracer_read_memory_state(const struct snap_device *dev){
+	smp_mb();
+	return atomic_read(&dev->sd_memory_state);
+}
+
+static inline void tracer_set_memory_state(struct snap_device *dev, int error){
+	smp_mb();
+	(void)atomic_cmpxchg(&dev->sd_memory_state, 0, error);
+	smp_mb();
```

Comment:
I'm not really sure why do we need memory barriers surrounding call of the `atomic_cmpxchg`. Are they really necessary? It seems that we already have some value about the memory state to set it here. And the atomic operation itself doesn't require barriers.

Reply:
Since memory barriers was used in `tracer_set_fail_state` I decided not to neglect them.

Reply:
Ah, I see. You've used this code as an example https://github.com/elastio/elastio-snap/blob/648239388a4cc2c35eb484c5938f1b1aae488677/src/elastio-snap.c#L1110-L1119

maybe author decided to play it safe or I don’t know something about some ancient Linux kernel versions )
Let it be then for the uniformity )